libpromises/mod_custom.c

```diff
@@ -0,0 +1,741 @@
+            "Promise with promiser '%s' was kept by promise module '%s'",
+            pp->promiser,
+            module->path);
+    }
+    else if (StringEqual(result_str, "not_kept"))
+    {
+        result = PROMISE_RESULT_FAIL;
+        cfPS(
+            ctx,
+            LOG_LEVEL_VERBOSE,
+            result,
+            pp,
+            &a,
+            "Promise with promiser '%s' was not kept by promise module '%s'",
+            pp->promiser,
+            module->path);
+    }
+    else if (StringEqual(result_str, "repaired"))
+    {
+        result = PROMISE_RESULT_CHANGE;
+        cfPS(
+            ctx,
+            LOG_LEVEL_VERBOSE,
+            result,
+            pp,
+            &a,
+            "Promise with promiser '%s' was repaired by promise module '%s'",
+            pp->promiser,
+            module->path);
+    }
+    else if (StringEqual(result_str, "error"))
+    {
+        result = PROMISE_RESULT_FAIL;
+        cfPS(
+            ctx,
+            LOG_LEVEL_ERR,
+            result,
+            pp,
+            &a,
+            "An unexpected error occured in promise module (bug in promise module: %s)",
+            module->path);
+    }
+    else
+    {
+        result = PROMISE_RESULT_FAIL;
+        cfPS(
+            ctx,
+            LOG_LEVEL_ERR,
+            result,
+            pp,
+            &a,
+            "Promise module returned unacceptable result: '%s' (bug in promise module: %s)",
+            result_str,
+            module->path);
+    }
+
+    JsonDestroy(response);
+    return result;
+}
+
+static void PromiseModule_Terminate(PromiseModule *module)
+{
+    if (module != NULL)
+    {
+        PromiseModule_AppendString(module, "operation", "terminate");
+        PromiseModule_Send(module);
+
+        JsonElement *response = PromiseModule_Receive(module);
+        JsonDestroy(response);
+
+        PromiseModule_DestroyInternal(module);
+    }
+}
+
+PromiseResult EvaluateCustomPromise(EvalContext *ctx, const Promise *pp)
+{
+    assert(ctx != NULL);
+    assert(pp != NULL);
+
+    Body *promise_block = FindCustomPromiseType(pp);
+    char *interpreter = NULL;
+    char *path = NULL;
+
+    bool success = GetInterpreterAndPath(promise_block, &interpreter, &path);
+
+    if (!success)
+    {
+        assert(interpreter == NULL && path == NULL);
+        return PROMISE_RESULT_FAIL;
+    }
+
+    // TODO: Store promise modules and avoid starting one per promise
+    // evaluation
+    PromiseModule *module = PromiseModule_Start(interpreter, path);
+    PromiseResult result;
+
+    // TODO: Do validation earlier (cf-promises --full-check)
+    bool valid = PromiseModule_Validate(module, pp);
+
+    if (valid)
+    {
+        valid &= CustomPromise_IsFullyResolved(pp);
+        if (!valid)
+        {
+            Log(LOG_LEVEL_ERR,
+                "%s promise with promiser '%s' has unresolved/unexpanded variables",
+                pp->parent_section->promise_type,
+                pp->promiser);
```

Comment:
I take it this doesn't result in `error: ` messages being printed before the variables are expanded and the promise is evaluated?